File: 2025/261. Graph Valid Tree/Solution.java
class Solution {
    public boolean validTree(int n, int[][] edges) {
        if (edges.length != n - 1)
            return false;
        Map<Integer, List<Integer>> map = new HashMap<>();
        for (int[] e : edges) {
            map.computeIfAbsent(e[0], k -> new ArrayList<>()).add(e[1]);
            map.computeIfAbsent(e[1], k -> new ArrayList<>()).add(e[0]);
        }
        Set<Integer> visited = new HashSet<>();
        Stack<Integer> stack = new Stack<>();
        stack.push(0);
        while (!stack.isEmpty()) {
            int node = stack.pop();
            if (visited.contains(node))
                continue;
            visited.add(node);
            if (map.containsKey(node)) {
                 for (int neighbor : map.get(node)) {
                    stack.push(neighbor);
                }
            }
        }
        return visited.size() == n;
    }
}

File: 2025/261. Graph Valid Tree/solution.py
class Solution:
    def validTree(self, n: int, edges: List[List[int]]) -> bool:
        if n - 1 != len(edges):
            return False

        neighbor_list = [[] for _ in range(n)]
        for origin, dest in edges:
            neighbor_list[origin].append(dest)
            neighbor_list[dest].append(origin)

        visited = set()
        def dfs_traverse(node: int):
            if node in visited: return
            visited.add(node)
            for n in neighbor_list[node]:
                dfs_traverse(n)
        dfs_traverse(0)
        return len(visited) == n
